File: Python/Array/Bag of Tokens(L 948).py
#Problem link:https://leetcode.com/problems/bag-of-tokens/

#Time Complexity O(N*log(N)) - N = tokens length
#Space Complexity O(1)

class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        
        #score - current score
        #max_score = maximum score
        score = max_score = 0 
        tokens.sort()
        queue = collections.deque(tokens)

        while queue:
            if power >= queue[0]:#if we have enough power, increase the score
                power -= queue.popleft()    # pop the smallest power from left
                score += 1
                max_score = max(max_score, score)#check if we have a new max score
            elif score >= 1:#if we don't have enough power, increase the power
                power += queue.pop()        # pop the biggest power from right
                score -= 1
            else:
                queue.popleft()
        
        return max_score

File: Java/Array/Bag of Tokens(L 948).java
// problem Link: https://leetcode.com/problems/bag-of-tokens/

/****
 * Approach: need to return maximum possible score
 * we gain score when our power is greater than tokens[i]
 * and can loose score if our power is less
 * 
 * first we sort array
 * and then iterate through tokens array, if our power is greater
 * than tokens[i], it means we can gain score by lossing ouw power.
 * 
 * but if our power is less than tokens[i], it means we can gain power by
 * tokens[i] that have more value.
 */


/****
 * Time Complexity: O(N logN), N=length of tokens array
 * Space Complexity: O(1)
 */


import java.util.Arrays;

class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int res = 0, points = 0, i = 0, j = tokens.length - 1;
        while (i <= j) {
            if (power >= tokens[i]) {
                power -= tokens[i++];
                res = Math.max(res, ++points);
            } else if (points > 0) {
                points--;
                power += tokens[j--];
            } else {
                break;
            }
        }
        return res;
    }
}